public/app/plugins/panel/table/models.gen.ts

```diff
@@ -0,0 +1,33 @@
+// Autogenerated from models.cue. DO NOT EDIT.
+// (actually this is a handwritten mockup with explanatory comments right now,
+// but it's representative!)
+
+import { TableCellDisplayMode, TableSortByFieldState } from '@grafana/ui';
+
+// Only the latest schema version is translated to TypeScript, on the premise
+// that either the dashboard loading process, or (eventually) CUE-defined
+// migrations ensure that bulk of the frontend application only ever
+// need directly consider the most recent version of the schema.
+export const modelVersion = Object.freeze([1, 0]);
+
```

Comment:
The changes in this folder are likely the most interesting for frontend people to explore.  The key idea is that `models.cue` defines what we want to save in panel JSON -- a process (still TBD, but based on cuetsy) will generate the `models.gen.ts` from that definition.

In `module.ts` we can reference the defaults directly -- this will help guarantee that the models and UI stay in sync

Reply:
This is an exciting proposal. I guess we will have to figure out what to do with the types now to be generated from models.cue that currently live in grafana/ui?

Reply:
Ah, nvm, I now see cue files in the packages from which the corresponding types will be exported. 

Question - should we assume the cue based types are public API? Wondering how should we approach the APIs of the packages with the generated types.

Reply:
Yes, the cue files do form a public API. The Cue files can be used to generate libraries in multiple languages as well as clear API docs.

Reply:
> should we assume the cue based types are public API? Wondering how should we approach the APIs of the packages with the generated types.

Yup, as Malcolm said, they do form a public API. And not just one, but three - in TypeScript, Go, _and_ CUE. These should obviously be equivalent, and they are...but with caveats. This file provides a good example:

* **CUE:** the scuemata family is declared directly in `models.cue`. All versions of the schema are accessible for external import. The file itself has a package on it (though `grafanschema` may not be what we ultimately want here), which means it's importable by CUE rules. And the presence of the `cue.mod` dir and module root means that this package has a canonical importable path - `import "github.com/grafana/grafana/public/app/plugins/panel/table:grafanaschema"`. But current thinking is that non-core plugins should not, and indeed cannot, have a `cue.mod` and module path in them, as it will break the interpretation of import statements that point to Grafana's shared components, as the table plugin imports here.
    * Interestingly, the full and correct representation of the scuemata actually _isn't_ available from CUE, and never quite will be. This is because it's impossible to dynamically import CUE from within a `.cue` file, and to fully express a dashboard schema, we have to compose from the sub-schema declared in the `models.cue` files of all plugins. Concretely: while we _could_ import the table's plugin's models.cue, and the text plugin, etc., and unify them into the root dashboard schema, we can't ever enumerate all possible contributed plugins - what about private ones? - in order to explicitly import them. Fortunately, this is only a blocker to being able to create a truly universal schema for all possible valid dashboards; for any particular CUE consumer of the schemas, they can know that they want to import, say, [worldmap](https://grafana.com/grafana/plugins/grafana-worldmap-panel/), import it in their own project (once we solve the make-plugins-importable problem) and compose it in themselves.
* **Go:** the full scuemata family will be made available via the [general Go interface](https://github.com/grafana/grafana/pull/32527/files#diff-e112952d4a4e1398163b57958ef00de86d89f769005526d7d7d1728de6e75ca0). Right now, though, there isn't an exported loader function to produce an instance of this interface. ([`load.readPanelModels()`](https://github.com/grafana/grafana/pull/32527/files#diff-f37c80b8b7621396f715bcfc15262d0c833166cc5f16f24734a38627c0866888R69) is the private func that does it rn.) That's a transitional state until i get composite schema fully genericized, at which point the successor to [`LatestPanelSchemaFor()`](https://github.com/grafana/grafana/pull/32527/files#diff-b31ceb59c6b295516d72429dcba1a127a0349e060b09523e8c5de43ea9c3d67fR199-R207) will allow the user to call up either a particular schema, or the entire family, of any sub-schema that's been composed into a composite schema. 
    * What we have right now is [the hack](https://github.com/grafana/grafana/pull/32527/files#diff-b31ceb59c6b295516d72429dcba1a127a0349e060b09523e8c5de43ea9c3d67fR199-R207) that allows calling up the latest schema from a family, but that is a) only the latest schema and b) does not expose the schema in its original, declared form from `models.cue`, but rather in its shape after it is [transformed](https://github.com/grafana/grafana/pull/32527/files#diff-f37c80b8b7621396f715bcfc15262d0c833166cc5f16f24734a38627c0866888R48-R62) into the shape that allows it to be [unified](https://github.com/grafana/grafana/pull/32527/files#diff-b31ceb59c6b295516d72429dcba1a127a0349e060b09523e8c5de43ea9c3d67fR94) with the base dashboard schema.
    * Note that the core and core-plugin CUE files are all embedded in the Go packages, which means that all the patterns and guarantees that are available to the Grafana backend are equally available to any other Go program that wants to import them.
* **TypeScript:** the TS representation of the scuemata family is in this file, `models.gen.ts`. Clearly, we only have the latest version of the schema here, rather than the whole family. That's by design, and we do not intend it to change; since there is no Grafana frontend app without a corresponding Grafana backend, we can delegate migration operations entirely to the backend, and the frontend app can pretend like the only schema that exists is the latest one. (That's all eventually, once the schema are mature enough for us to fully rely on them.)

So yes, each language presents an API for interacting with these schema, and where the same schema@version is in play, the actual schemas should be equivalent, _but_, not all versions of all schema will be available in all language contexts, and the schema-based operations supported vary as well.

There's some nuance here, though, and it's one of the TODOs (which i have to post up more thoroughly): right now, different parts of the system are

Reply:
@dprokop -- your question gets to the core issue we have.  Currently in packages there is a mix of "interfaces needed for components" and "stuff saved in panel json".  The goal here is to be very clear about which is required for what.  We currently do this haphazardly and then need to reverse engineer it.  In this proposal, that will be explicit by putting stuff saved in json in the 'models.cue' file